sys/net/netstats/netstats_neighbor.c

```diff
@@ -0,0 +1,421 @@
+/*
+ * Copyright (C) Koen Zandberg <koen@bergzand.net>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net
+ * @file
+ * @brief       Neighbor level stats for netdev
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ * @author      Benjamin Valentin <benpicco@beuth-hochschule.de>
+ * @}
+ */
+
+#include <errno.h>
+
+#include "net/netdev.h"
+#include "net/netstats/neighbor.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static inline void _lock(netif_t *dev)
+{
+    mutex_lock(&dev->neighbors.lock);
+}
+
+static inline void _unlock(netif_t *dev)
+{
+    mutex_unlock(&dev->neighbors.lock);
+}
+
+/**
+ * @brief Compare the freshness of two records
+ *
+ * @param[in] a     pointer to the first record
+ * @param[in] b     pointer to the second record
+ * @param[in] now   current timestamp in seconds
+ *
+ * @return       pointer to the least fresh record
+ */
+static inline netstats_nb_t *netstats_nb_comp(const netstats_nb_t *a,
+                                              const netstats_nb_t *b,
+                                              uint16_t now)
+{
+    return (netstats_nb_t *)((now - a->last_updated > now - b->last_updated) ? a : b);
+}
+
+static bool l2_addr_equal(const uint8_t *a, uint8_t a_len, const uint8_t *b, uint8_t b_len)
+{
+    if (a_len != b_len) {
+        return false;
+    }
+
+    return memcmp(a, b, a_len) == 0;
+}
+
+static void netstats_nb_half_freshness(netstats_nb_t *stats, timex_t *cur)
+{
+    uint16_t now = cur->seconds;
+    uint8_t diff = (now - stats->last_halved) / NETSTATS_NB_FRESHNESS_HALF;
+    stats->freshness >>= diff;
+
+    if (diff) {
+        /* Set to the last time point where this should have been halved */
+        stats->last_halved = now - diff;
+    }
+}
+
+static void netstats_nb_incr_freshness(netstats_nb_t *stats)
+{
+    timex_t cur;
+    xtimer_now_timex(&cur);
+
+    /* First halve the freshness if applicable */
+    netstats_nb_half_freshness(stats, &cur);
+
+    /* Increment the freshness capped at FRESHNESS_MAX */
+    if (stats->freshness < NETSTATS_NB_FRESHNESS_MAX) {
+        stats->freshness++;
+    }
+
+    stats->last_updated = cur.seconds;
+}
+
+bool netstats_nb_isfresh(netstats_nb_t *stats)
+{
+    timex_t cur;
+    xtimer_now_timex(&cur);
+    uint16_t now = cur.seconds;
+
+    /* Half freshness if applicable to update to current freshness */
+    netstats_nb_half_freshness(stats, &cur);
+
+    return (stats->freshness >= NETSTATS_NB_FRESHNESS_TARGET) &&
+           (now - stats->last_updated < NETSTATS_NB_FRESHNESS_EXPIRATION);
+}
+
+void netstats_nb_init(netif_t *dev)
+{
+    mutex_init(&dev->neighbors.lock);
+    memset(dev->neighbors.pstats, 0, sizeof(netstats_nb_t) * NETSTATS_NB_SIZE);
+    cib_init(&dev->neighbors.stats_idx, NETSTATS_NB_QUEUE_SIZE);
```

Comment:
Lock mutex around this?

Reply:
already locked by `netstats_nb_record()`

Reply:
But `netstats_nb_record()` does not call `netstats_nb_init()`??!? Nevertheless, this is an exposed function, so it should enter the mutex, ideally. Usually a deadlock that is caused by that shows either, that something is wrong or an `rmutex` should have been used instead.

Reply:
Ah wrong line - but this is only called by `gnrc_netif_create()`.
I can lock it regardless.